### DISCUSSION TITLE
Polygon line edge is stepping stairs

### DISCUSSION THREAD

**Discussion by gmek5476** [ImageMagick/ImageMagick#6556] [General]
Hi All,
we used the below command to draw a ploygon on the image ,But after drawing the polygon we noticed tht the diagonal line edge is appears like a stepping stairs .Is there anyway to have like a smooth line instead of stepping stairs.

magick C:\Im-Test\Input\pexels-photo-3028127-S.jpg   ( +clone -fill white -colorize 100 -fill black +antialias  -draw " polygon 0,0 0,%[fx:h-1] %[fx:(35/100)*w],%[fx:h-1] %[fx:((35/100)*w)+h*tan(pi*(90-70)/180)],0" ) -alpha off -compose CopyOpacity -composite -compose over -background yellow  -flatten C:\Im-Test\Output\cut-Out\issue12.jpg

Please find the images below
![pexels-photo-3028127-S](https://github.com/ImageMagick/ImageMagick/assets/126016862/a577c562-ccda-49b6-8747-39f6d6e33e9e)
![issue12](https://github.com/ImageMagick/ImageMagick/assets/126016862/9103cfec-88af-4553-a679-fda443f0304b)
<img width="184" alt="Capture" src="https://github.com/ImageMagick/ImageMagick/assets/126016862/d7089807-7f0c-4f5c-8185-ddf7ac447314">
![issue12](https://github.com/ImageMagick/ImageMagick/assets/126016862/a6739ba9-edd7-4d14-ac67-b7e7d8656bc6)

Any help is really appreciated.

**Comment by fmw42** [comment]
If you want a smoother line, remove +antialias.  That actually aliases the line.

See definition:  "Use +antialias to disable the addition of antialiasing edge pixels.”

Fred

**Reply by gmek5476** [reply]
Hi fred initally i tried with out +antialias it is also getting the same stepping stairs line. Please find the command below
magick C:\Im-Test\Input\pexels-photo-3028127-S.jpg   ( +clone -fill white -colorize 100 -fill black  -draw " polygon 0,0 0,%[fx:h-1] %[fx:(35/100)*w],%[fx:h-1] %[fx:((35/100)*w)+h*tan(pi*(90-70)/180)],0" ) -alpha off -compose CopyOpacity -composite -compose over -background yellow  -flatten C:\Im-Test\Output\cut-Out\issue12.jpg

![issue12](https://github.com/ImageMagick/ImageMagick/assets/126016862/22bab7dd-cd1c-4b97-a6f0-880dc324cedf)
![issue12](https://github.com/ImageMagick/ImageMagick/assets/126016862/a317c5ff-cd29-486e-91ff-58dcad62e1eb)

**Comment by snibgo** [comment]
@gmek5476: Your command has become mangled by the forum software. Please show it again, but this time insert a line of three backticks above it, and the same below it. Three backticks are like this: \`\`\`

**Reply by gmek5476** [reply]
Here is the Command
```magick C:\Im-Test\Input\pexels-photo-3028127-S.jpg   ( +clone -fill white -colorize 100 -fill black  -draw " polygon 0,0 0,%[fx:h-1] %[fx:(35/100)*w],%[fx:h-1] %[fx:((35/100)*w)+h*tan(pi*(90-70)/180)],0" ) -alpha off -compose CopyOpacity -composite -compose over -background yellow  -flatten C:\Im-Test\Output\cut-Out\issue12.jpg```

**Comment by fmw42** [comment]
I do not see anything wrong.  How did you zoom in?  If you zoom with a tool that just replicates pixels that will introduce stair-stepping.  You cannot look any further than full resolution, that is, just crop at full resolution, do not zoom.

Furthermore, there is only so much one can do to anti-alias with the -antialias command. The rest you must do with a blur.

Here are my Unix commands and the results.

```
magick photo.jpg \( +clone -fill white -colorize 100 -fill black +antialias -draw "polygon 0,0 0,%[fx:h-1] %[fx:(35/100)*w],%[fx:h-1] %[fx:((35/100)*w)+h*tan(pi*(90-70)/180)],0" \) -alpha off -compose CopyOpacity -composite -compose over -background yellow -flatten photo_proc_alias.png

magick photo.jpg \( +clone -fill white -colorize 100 -fill black -antialias -draw "polygon 0,0 0,%[fx:h-1] %[fx:(35/100)*w],%[fx:h-1] %[fx:((35/100)*w)+h*tan(pi*(90-70)/180)],0" \) -alpha off -compose CopyOpacity -composite -compose over -background yellow -flatten photo_proc_antialias.png

magick photo.jpg \( +clone -fill white -colorize 100 -fill black -antialias -draw "polygon 0,0 0,%[fx:h-1] %[fx:(35/100)*w],%[fx:h-1] %[fx:((35/100)*w)+h*tan(pi*(90-70)/180)],0" -blur 0x2 -level 40x60% \) -alpha off -compose CopyOpacity -composite -compose over -background yellow -flatten photo_proc_antialias2.png
```

I note that you are missing your * for multiply. Perhaps post lost them or in the email.


Here are my results. I saved as PNG to avoid JPG artifacts.  The last has blur for more anti-aliasing. If not enough, increase the -blur amount.

![photo_proc_alias](https://github.com/ImageMagick/ImageMagick/assets/13458218/5658a165-c8be-4afb-a4b5-83bb12e1dcb6)

![photo_proc_antialias](https://github.com/ImageMagick/ImageMagick/assets/13458218/a834c6cc-2830-4221-a04b-930d758fc418)

![photo_proc_antialias2](https://github.com/ImageMagick/ImageMagick/assets/13458218/5129d0f7-0396-4c24-8110-3a104b46a8b4)

**Comment by snibgo** [comment]
The images in the first post are not anti-aliased, so they have visible "stair-casing". This is because of "`+antialias`".

The image in the later post, with no "`+antialias`", have less obvious stair-casing. If we zoom in, we can see that yellow has various densities. and the anti-aliasing has worked.

Yellow versus black is high contrast. If the screen and eysight are good enough to see individual pixels, some stair-casing will be visible.

As fmw42 says, blurring will reduce the effect.

Another possibility is to do the flattening in linear RGB space. This may help. Change...
```
-flatten
```
... to:
```
-colorspace RGB -flatten -colorspace sRGB
```

**Comment by snibgo** [comment]
Here is an example of flattening in sRGB and linear RGB spaces (Windows BAT syntax).
```
magick ^
  -size 300x300 ^
  xc:Black -write mpr:BLK +delete ^
  xc:None -fill Yellow -stroke None ^
  -draw "polygon 0,0 150,0 100,299 0,299" ^
  -write mpr:YEL +delete ^
  ( mpr:BLK mpr:YEL -layers flatten ) ^
  ( mpr:BLK mpr:YEL -colorspace RGB -layers flatten -colorspace sRGB ) ^
  -bordercolor gray(50%%) -border 20 ^
  +append +repage ^
  x.png
```
The result is:

![x](https://github.com/ImageMagick/ImageMagick/assets/33812368/ff874db2-5b76-402b-9a93-21cd88a5c031)

The two halves have the same anti-aliasing. To my eyes, on my computer, the left side (flattened in sRGB space) has more noticeable jaggies than the right side (flattened in linear RGB space). 

Any operation that combines two or more pixels to make each output pixel will look different depending on the colorspace used. The difference is more noticeable in graphics (bold colours, sharp edges) than in photos (dull colours, blurred edges).
